First empty line in code fragment breaks line numeration in error messages

Test doc:

<pre>
### Heading

```js

console.log('a')
```
</pre>

Message:

```console
  4:17  error  Missing semicolon  semi
```

This should be `5:17`.

--- a/package.json
+++ b/package.json
@@ -47,7 +47,7 @@
     "nyc": "^14.1.1"
   },
   "dependencies": {
-    "remark-parse": "^5.0.0",
+    "remark-parse": "^7.0.0",
     "unified": "^6.1.2"
   },
   "peerDependencies": {

--- a/tests/lib/plugin.js
+++ b/tests/lib/plugin.js
@@ -145,6 +145,21 @@ describe("plugin", () => {
         assert.strictEqual(report.results[0].messages[1].endLine, 8);
     });
 
+    // https://github.com/eslint/eslint-plugin-markdown/issues/77
+    it("should emit correct line numbers with leading blank line", () => {
+        const code = [
+            "### Heading",
+            "",
+            "```js",
+            "",
+            "console.log('a')",
+            "```"
+        ].join("\n");
+        const report = cli.executeOnText(code, "test.md");
+
+        assert.strictEqual(report.results[0].messages[0].line, 5);
+    });
+
     it("doesn't add end locations to messages without them", () => {
         const code = [
             "```js",
@@ -277,6 +292,23 @@ describe("plugin", () => {
             assert.strictEqual(report.results[0].messages[3].line, 15);
         });
 
+        // https://github.com/eslint/eslint-plugin-markdown/issues/78
+        it("preserves leading empty lines", () => {
+            const code = [
+                "<!-- eslint lines-around-directive: ['error', 'never'] -->",
+                "",
+                "```js",
+                "",
+                "\"use strict\";",
+                "```"
+            ].join("\n");
+            const report = cli.executeOnText(code, "test.md");
+
+            assert.strictEqual(report.results.length, 1);
+            assert.strictEqual(report.results[0].messages.length, 1);
+            assert.strictEqual(report.results[0].messages[0].message, "Unexpected newline before \"use strict\" directive.");
+            assert.strictEqual(report.results[0].messages[0].line, 5);
+        });
     });
 
     describe("should fix code", () => {

--- a/tests/lib/processor.js
+++ b/tests/lib/processor.js
@@ -199,7 +199,21 @@ describe("processor", () => {
 
             assert.strictEqual(blocks.length, 1);
             assert.strictEqual(blocks[0].filename, "0.js");
-            assert.strictEqual(blocks[0].text, "\n\n \n  \n");
+            assert.strictEqual(blocks[0].text, "\n\n\n \n  \n");
+        });
+
+        it("should preserve leading and trailing empty lines", () => {
+            const code = [
+                "```js",
+                "",
+                "console.log(42);",
+                "",
+                "```"
+            ].join("\n");
+            const blocks = processor.preprocess(code);
+
+            assert.strictEqual(blocks[0].filename, "0.js");
+            assert.strictEqual(blocks[0].text, "\nconsole.log(42);\n\n");
         });
 
         it("should ignore code fences with unspecified info string", () => {

